src/persist-client/src/internal/compact.rs

```diff
@@ -468,34 +468,76 @@ where
 
         let mut timings = Timings::default();
 
-        // populate our heap with the updates from the first part of each run
-        for (index, (part_desc, parts)) in runs.iter_mut().enumerate() {
+        // populate our heap with the updates from the first part of each run.
+        // we have the memory to fetch each part in parallel, so we spawn a
+        // task to fetch each one.
+        let mut pending_initial_parts = VecDeque::with_capacity(usize::min(
+            runs.len(),
+            cfg.compaction_reads_max_outstanding_parts,
+        ));
+        let mut fetched_initial_parts = Vec::with_capacity(runs.len());
+        for (runs_index, (part_desc, parts)) in runs.iter_mut().enumerate() {
             if let Some(part) = parts.next() {
-                let start = Instant::now();
-                let mut part = fetch_batch_part(
-                    shard_id,
-                    blob.as_ref(),
-                    &metrics,
-                    &metrics.read.compaction,
-                    &part.key,
-                    part_desc,
-                )
-                .await?;
-                timings.part_fetching += start.elapsed();
-                let start = Instant::now();
-                while let Some((k, v, mut t, d)) = part.next() {
-                    t.advance_by(desc.since().borrow());
-                    let d = D::decode(d);
-                    let k = k.to_vec();
-                    let v = v.to_vec();
-                    // default heap ordering is descending
-                    sorted_updates.push(Reverse((((k, v), t, d), index)));
-                    remaining_updates_by_run[index] += 1;
+                let shard_id = shard_id.clone();
+                let blob = Arc::clone(&blob);
+                let metrics = Arc::clone(&metrics);
+                let compaction_metrics = metrics.read.compaction.clone();
+                let key = part.key.clone();
+                let part_desc = (**part_desc).clone();
+
+                pending_initial_parts.push_back((
+                    runs_index,
+                    mz_ore::task::spawn(|| "compaction::initial_fetch_batch_part", async move {
+                        fetch_batch_part(
+                            &shard_id,
+                            blob.as_ref(),
+                            &metrics,
+                            &compaction_metrics,
+                            &key,
+                            &part_desc,
+                        )
+                        .await
+                    }),
+                ));
+
+                if pending_initial_parts.len() > cfg.compaction_reads_max_outstanding_parts {
+                    let start = Instant::now();
+                    let (index, part) = pending_initial_parts
+                        .pop_front()
```

Comment:
We always pop off the front, and not necessarily the next ready part. I know we're doing it because we want to preserve the run order but in bad cases, say if the largest part sits at the front while all the others are already finished we're loosing parallelization potential.

Could we, instead, just sort `fetched_initial_parts` by the "run index"? 

Also: I think this is fine as is, so feel free to merge. :+1:

Reply:
Yea this model definitely has some pathologies like you mention. Any ideas on maintaining the concurrency limit otherwise?

Reply:
I think for this PR I'll keep it as-is (it's the same model we use for in-flight writes to S3), but when we eventually start eagerly pulling down the rest of the blobs, we can update how this works.